File: scripts/lib/gitmeta.py
"""Defensive git facts for the analyzer + incremental build.

Every function shells out to `git` with cwd=root and returns a safe empty
value on ANY failure (git absent, not a repo, command error) — it must never
raise into the pipeline. Incremental build is a pure optimization; when git is
unavailable the caller falls back to a full build.
"""
from __future__ import annotations

import subprocess
from pathlib import Path


def _run(args, root):
    """Run a git command under `root`; None if git is missing/timed out."""
    try:
        return subprocess.run(["git", *args], cwd=str(root),
                              capture_output=True, text=True, timeout=15)
    except (FileNotFoundError, subprocess.TimeoutExpired):
        return None


def is_git_repo(root: Path) -> bool:
    cp = _run(["rev-parse", "--is-inside-work-tree"], root)
    return bool(cp and cp.returncode == 0 and cp.stdout.strip() == "true")


def toplevel(root: Path):
    cp = _run(["rev-parse", "--show-toplevel"], root)
    return cp.stdout.strip() if cp and cp.returncode == 0 else None


def git_info(root: Path):
    """branch / commit / short / dirty for HEAD. None when not a repo or HEAD
    can't resolve (e.g. a repo with no commits yet)."""
    if not is_git_repo(root):
        return None
    head = _run(["rev-parse", "HEAD"], root)
    if not head or head.returncode != 0:
        return None
    commit = head.stdout.strip()
    br = _run(["rev-parse", "--abbrev-ref", "HEAD"], root)
    branch = br.stdout.strip() if br and br.returncode == 0 else "HEAD"
    st = _run(["status", "--porcelain"], root)
    dirty = bool(st and st.returncode == 0 and st.stdout.strip())
    return {"branch": branch, "commit": commit, "short": commit[:7], "dirty": dirty}


def is_ancestor(root: Path, base: str) -> bool:
    """True iff `base` is reachable from HEAD."""
    cp = _run(["merge-base", "--is-ancestor", base, "HEAD"], root)
    return bool(cp and cp.returncode == 0)


def changed_files(root: Path, base: str):
    """Rel paths changed between `base` and the working tree: union of
    committed diff (base..HEAD) and worktree (staged+unstaged+untracked).
    None if git is unusable."""
    diff = _run(["diff", "--name-only", base, "HEAD"], root)
    if not diff or diff.returncode != 0:
        return None
    out = {line.strip() for line in diff.stdout.splitlines() if line.strip()}
    st = _run(["status", "--porcelain"], root)
    if st and st.returncode == 0:
        for line in st.stdout.splitlines():
            if not line.strip():
                continue
            rest = line[3:] if len(line) > 3 else line.strip()
            if " -> " in rest:                 # rename: "old -> new"
                rest = rest.split(" -> ", 1)[1]
            rest = rest.strip().strip('"')
            if rest:
                out.add(rest)
    return out

File: scripts/lib/gitmeta.mjs
// scripts/lib/gitmeta.mjs — defensive git facts. Port of gitmeta.py.
// Every function returns a safe empty value on ANY failure (git absent, not a
// repo, command error) — it must never throw into the pipeline.
import { execFileSync } from 'node:child_process';

function run(args, root) {
  try {
    const stdout = execFileSync('git', args, {
      cwd: String(root), encoding: 'utf8', timeout: 15000,
      stdio: ['ignore', 'pipe', 'ignore'],
    });
    return { code: 0, stdout };
  } catch (e) {
    // non-zero exit still carries stdout; missing git / timeout → null
    if (e && e.stdout != null && e.status != null) return { code: e.status, stdout: String(e.stdout) };
    return null;
  }
}

export function isGitRepo(root) {
  const cp = run(['rev-parse', '--is-inside-work-tree'], root);
  return !!(cp && cp.code === 0 && cp.stdout.trim() === 'true');
}

export function toplevel(root) {
  const cp = run(['rev-parse', '--show-toplevel'], root);
  return cp && cp.code === 0 ? cp.stdout.trim() : null;
}

export function gitInfo(root) {
  if (!isGitRepo(root)) return null;
  const head = run(['rev-parse', 'HEAD'], root);
  if (!head || head.code !== 0) return null;
  const commit = head.stdout.trim();
  const br = run(['rev-parse', '--abbrev-ref', 'HEAD'], root);
  const branch = br && br.code === 0 ? br.stdout.trim() : 'HEAD';
  const st = run(['status', '--porcelain'], root);
  const dirty = !!(st && st.code === 0 && st.stdout.trim());
  return { branch, commit, short: commit.slice(0, 7), dirty };
}

export function isAncestor(root, base) {
  const cp = run(['merge-base', '--is-ancestor', base, 'HEAD'], root);
  return !!(cp && cp.code === 0);
}

export function changedFiles(root, base) {
  const diff = run(['diff', '--name-only', base, 'HEAD'], root);
  if (!diff || diff.code !== 0) return null;
  const out = new Set(diff.stdout.split('\n').map((l) => l.trim()).filter(Boolean));
  const st = run(['status', '--porcelain'], root);
  if (st && st.code === 0) {
    for (const line of st.stdout.split('\n')) {
      if (!line.trim()) continue;
      let rest = line.length > 3 ? line.slice(3) : line.trim();
      if (rest.includes(' -> ')) rest = rest.split(' -> ')[1];
      rest = rest.trim().replace(/^"|"$/g, '');
      if (rest) out.add(rest);
    }
  }
  return out;
}
